src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1283,15 +1329,42 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
             BinaryPatternSyntax node,
             TypeSymbol inputType,
             uint inputValEscape,
+            out TypeSymbol outputType,
+            out uint outputValEscape,
             bool permitDesignations,
             bool hasErrors,
             DiagnosticBag diagnostics)
         {
             bool isDisjunction = node.PatternOperator.Kind() == SyntaxKind.OrKeyword;
-            permitDesignations = permitDesignations & !isDisjunction; // prevent designators under 'or'
-            var left = BindPattern(node.LeftPattern, inputType, inputValEscape, permitDesignations, hasErrors, diagnostics);
-            var right = BindPattern(node.RightPattern, inputType, inputValEscape, permitDesignations, hasErrors, diagnostics);
-            return new BoundBinaryPattern(node, disjunction: isDisjunction, left, right, inputType, hasErrors);
+            if (isDisjunction)
+            {
+                permitDesignations = false; // prevent designators under 'or'
+                var left = BindPattern(node.LeftPattern, inputType, inputValEscape, out var leftOutputType, out var leftOutputValEscape, permitDesignations, hasErrors, diagnostics);
+                var right = BindPattern(node.RightPattern, inputType, inputValEscape, out var rightOutputType, out var rightOutputValEscape, permitDesignations, hasErrors, diagnostics);
+                outputType = commonType(node, leftOutputType, rightOutputType, diagnostics) ?? inputType;
+                outputValEscape = GetValEscape(outputType, inputValEscape);
+                return new BoundBinaryPattern(node, disjunction: isDisjunction, left, right, outputType: outputType, inputType: inputType, hasErrors);
+            }
+            else
+            {
+                var left = BindPattern(node.LeftPattern, inputType, inputValEscape, out var leftOutputType, out var leftOutputValEscape, permitDesignations, hasErrors, diagnostics);
+                var right = BindPattern(node.RightPattern, leftOutputType, leftOutputValEscape, out outputType, out outputValEscape, permitDesignations, hasErrors, diagnostics);
+                return new BoundBinaryPattern(node, disjunction: isDisjunction, left, right, outputType: outputType, inputType: inputType, hasErrors);
+            }
+
+            TypeSymbol commonType(SyntaxNode node, TypeSymbol t1, TypeSymbol t2, DiagnosticBag diagnostics)
+            {
+                if (t1.Equals(t2, TypeCompareKind.AllIgnoreOptions))
```

Comment:
I would have expected a tighter check. What happens if we have two types differing in nullability? Should we take the first one or the second one, or should it depend on relational pattern (`or` or `and`)?

Reply:
doesn't this handle the issue your raised with `int or long` (`long` as the common type and convert ints), in all the other places we infer long for `2` and `2L` for example, but perhaps that's because we're relying on conversion-from-expression rather than from type?

Reply:
These and related questions are queued for LDM discussion (see 2020 agenda)

---
In reply to: [389183558](https://github.com/dotnet/roslyn/pull/42207#discussion_r389183558) [](ancestors = 389183558)

Reply:
These and related questions are queued for LDM discussion (see 2020 agenda)

---
In reply to: [389345455](https://github.com/dotnet/roslyn/pull/42207#discussion_r389345455) [](ancestors = 389345455)